### PR TITLE
tests implemented, Travis CI implemented, HttpParams creation fixed

### PR DESCRIPTION
**Pull request by fazpu**

### CODE REVIEW

File: demo-app/client/src/tests/patch/patch-order.service.spec.ts [hunk 1]
```diff
@@ -0,0 +1,44 @@
+// TODO: make the test case work
```
Comment (fazpu) [inline demo-app/client/src/tests/patch/patch-order.service.spec.ts]: The test was not passing. Will be fixed in the next PR. It is connected to how bodyParams are created.

File: demo-app/client/src/tests/put/put-order.service.spec.ts [hunk 2]
```diff
@@ -0,0 +1,47 @@
+// TODO: make the test case work
```
Comment (fazpu) [inline demo-app/client/src/tests/put/put-order.service.spec.ts]: the same as with patch-order.service.ts

File: src/requests/process-method.ts [hunk 3]
```diff
@@ -71,8 +71,17 @@ function getParamSeparation(paramGroups: Dictionary<Parameter[]>): string[] {
     let def: string;
 
     if (groupName === 'query') {
-      def = 'const queryParams = new HttpParams();\n';
-      def += _.map(group, p => `queryParams.append('${p.name}', JSON.stringify(params.${p.name}));`).join('\n');
+      def = 'const queryParams = new HttpParams()\n';
```
Comment (fazpu) [inline src/requests/process-method.ts]: the HttpParams creation did not work correctly before

**Comment by fazpu** [comment]
After the PR is merged, it will be necessary to activate the builds on Travis CI for this repo.

File: README.md [hunk 4]
```diff
@@ -37,7 +37,7 @@ The resulting API layer contains the following structure in the destination dire
 
 1. `def` directory stores all response interfaces and enums
 1. `model.ts` file reexports all of them together for a simple access
-1. `api` directory stores services devided by controllers containing all API methods
```
Comment (jnwltr) [inline README.md]: Please, keep the information about how services are organized present.

File: README.md [hunk 5]
```diff
@@ -94,15 +94,37 @@ export class MyComponent implements OnInit {
...
+
 ### Docker image
 
 1. `docker build . -t swagger-angular-generator`
 1. `docker run -u $(id -u) -it -v "$PWD":/code swagger-angular-generator bash`
 1. `npm i`
 
+### Testing
+
+##### How the testing works:
```
Comment (jnwltr) [inline README.md]: We do not need a colon here (not used above). H4 instead of H5?

File: README.md [hunk 6]
```diff
@@ -94,15 +94,37 @@ export class MyComponent implements OnInit {
...
+1. `cd demo-app/client`
+1. `npm run generate-from-swagger`
+1. `npm run test`
+
+or instead of step 2 and 3 run: `npm run testci`
+
 ---
 
-### _Pull requests are welcome!_
+### _Pull requests are welcomed!_
```
Comment (jnwltr) [inline README.md]: No need for a change, *sth is welcome (adj)* if both correct and more frequent.

File: demo-app/client/.editorconfig [hunk 7]
```diff
@@ -0,0 +1,13 @@
+# Editor configuration, see http://editorconfig.org
```
Comment (jnwltr) [inline demo-app/client/.editorconfig]: Is this file needed, i.e. do we want to keep the settings of the demo app independent?

File: demo-app/client/README.md [hunk 8]
```diff
@@ -0,0 +1,27 @@
+# Client
```
Comment (jnwltr) [inline demo-app/client/README.md]: What do you think about removing files we do not actively use? E.g. does this one bring any value? I do not have strict opinion, but in general prefer minimalistic approach.

File: demo-app/client/src/app/app.component.html [hunk 9]
```diff
@@ -0,0 +1,20 @@
+<!--The content below is only a placeholder and can be replaced.-->
```
Comment (jnwltr) [inline demo-app/client/src/app/app.component.html]: Let's simplify this.

File: src/requests/process-method.ts (hunk 3 above)
Comment (jnwltr) [reply src/requests/process-method.ts]: Great. I am really disappointed by how HttpParams work comparing to original Http service.

File: src/requests/process-method.ts [hunk 10]
```diff
@@ -71,8 +71,17 @@ function getParamSeparation(paramGroups: Dictionary<Parameter[]>): string[] {
...
+      def += _.map(group,
+              p => {
+          if (p.type === 'string') {
+            return `.set('${p.name}', params.${p.name})`;
+          } else {
+            return `.set('${p.name}', JSON.stringify(params.${p.name}))`;
+          }
+        },
+      ).join('\n');
```
Comment (jnwltr) [inline src/requests/process-method.ts]: Not sure, are `.set` invocations indented to proper level?

File: demo-app/client/generated/controllers/DeleteOrder.ts [hunk 11]
```diff
@@ -0,0 +1,34 @@
+/* tslint:disable:max-line-length */
```
Comment (jnwltr) [inline demo-app/client/generated/controllers/DeleteOrder.ts]: Should this folder be present at all? I would remove it from the repo.

File: src/requests/process-method.ts (hunk 10 above)
Comment (jnwltr) [reply src/requests/process-method.ts]: I see it's not. There's also one thing more important than sparing `JSON.stringify` for string types (btw, I would test real type, not swagger type), not sending undefined params `?param1=undefined&param2=undefined`, so I suggest the following `if` block:

```typescript
if (groupName === 'query') {
  def = 'let queryParams = new HttpParams();\n';
  group.forEach(p => {
    def +=
    `if (params.${p.name} !== undefined) ` +
    `queryParams = queryParams.set('${p.name}', JSON.stringify(params.${p.name}));\n`;
  });

  return def;
}
```

File: README.md (hunk 5 above)
Comment (fazpu) [reply README.md]: done

File: demo-app/client/.editorconfig (hunk 7 above)
Comment (fazpu) [reply demo-app/client/.editorconfig]: removed

File: demo-app/client/README.md (hunk 8 above)
Comment (fazpu) [reply demo-app/client/README.md]: removed

File: demo-app/client/generated/controllers/DeleteOrder.ts (hunk 11 above)
Comment (fazpu) [reply demo-app/client/generated/controllers/DeleteOrder.ts]: I believe new potential users might find it useful to see the generated files so that they can get a better sense of how the system works and how the generated files look like.

File: src/requests/process-method.ts (hunk 10 above)
Comment (fazpu) [reply src/requests/process-method.ts]: JSON.stringify must not be used on string, otherwise the resulting query param look like this `stringField=%22example-producer%22` instead of `stringField=example-producer`, i.e. the quote signs are added.

File: README.md (hunk 6 above)
Comment (jnwltr) [reply README.md]: Me: _would you say Pull requests are welcomed, or just Pull requests are welcome? Both correct, or just one of them?_
Native brit: _both sound ok to me. (Maybe I'd choose the second, but a quick search of Google as corpus has lots of results for both.)_

File: demo-app/client/generated/controllers/DeleteOrder.ts (hunk 11 above)
Comment (jnwltr) [reply demo-app/client/generated/controllers/DeleteOrder.ts]: Makes sense. Re-generation should be done in pre-publish stage, right?

File: src/requests/process-method.ts [hunk 12]
```diff
@@ -68,11 +68,26 @@ export function processMethod(method: ControllerMethod): MethodOutput {
...
+
+      def += `Object.entries(queryParamBase).forEach(([key, value]) => {\n`;
+      def += indent(`if (value !== undefined) {\n`);
+      def += indent(indent(`if (typeof value === 'string') {\n`));
+      def += indent(indent(indent(`queryParams = queryParams.set(key, value);\n`)));
+      def += indent(indent(`} else {\n`));
+      def += indent(indent(indent(`queryParams = queryParams.set(key, JSON.stringify(value));\n`)));
+      def += indent(indent(`}\n`));
+      def += indent(`}\n`);
+      def += `});\n`;
```
Comment (fazpu) [inline src/requests/process-method.ts]: Now the HttpParams are created without `undefined` values.

**Comment by jnwltr** [comment]
Adding one more commit before merge in https://github.com/jnwltr/swagger-angular-generator/pull/3
